File: MB-Chat/review/review-escalation.ts
import { DEFAULT_REVIEW_POLICY } from "./review-policy";
import { REVIEW_REASONS } from "./review-reasons";
import type { ConfidenceGateInput, EscalationLevel, ReviewEscalation, ReviewPolicy } from "./types";

export function buildReviewEscalation(input: ConfidenceGateInput, reasons: string[], policy: ReviewPolicy = DEFAULT_REVIEW_POLICY): ReviewEscalation {
  let escalation_level: EscalationLevel = "routine";
  let requires_specialist = false;
  let auto_blocked = false;

  if (policy.specialistRiskLevels.includes(input.risk_level) || input.target_specialty) {
    escalation_level = "specialist";
    requires_specialist = true;
  }

  if (input.risk_level === "critical" || reasons.includes(REVIEW_REASONS.HALLUCINATION_RISK)) {
    escalation_level = "urgent";
  }

  if (reasons.includes(REVIEW_REASONS.POLICY_BLOCK)) {
    escalation_level = "blocked";
    auto_blocked = true;
  }

  return {
    escalation_level,
    escalation_reason: reasons,
    target_specialty: input.target_specialty,
    requires_specialist,
    auto_blocked,
  };
}

File: MetaBrain/review_py/review_escalation.py
from . import review_reasons
from .review_policy import DEFAULT_REVIEW_POLICY
from .types import ConfidenceGateInput, ReviewEscalation, ReviewPolicy


def build_review_escalation(
    item: ConfidenceGateInput,
    reasons: list[str],
    policy: ReviewPolicy = DEFAULT_REVIEW_POLICY,
) -> ReviewEscalation:
    escalation_level = "routine"
    requires_specialist = False
    auto_blocked = False

    if item.risk_level in policy.specialist_risk_levels or item.target_specialty:
        escalation_level = "specialist"
        requires_specialist = True

    if item.risk_level == "critical" or review_reasons.HALLUCINATION_RISK in reasons:
        escalation_level = "urgent"

    if review_reasons.POLICY_BLOCK in reasons:
        escalation_level = "blocked"
        auto_blocked = True

    return ReviewEscalation(
        escalation_level=escalation_level,  # type: ignore[arg-type]
        escalation_reason=reasons,
        target_specialty=item.target_specialty,
        requires_specialist=requires_specialist,
        auto_blocked=auto_blocked,
    )
